### PR TITLE
 Update helium-typescript to use async/await pattern

### LINKED ISSUE #78: Update helium-typescript to use async/await pattern
# Design Review

## Introduction
- Use of promises vs async/await is inconsistent throughout and should be standardized for a better developer experience 
- Review all async methods to make sure are rejects are handled for future proofing
     -  [DEP0018: Unhandled promise rejections](https://nodejs.org/dist/latest-v8.x/docs/api/deprecations.html#deprecations_dep0018_unhandled_promise_rejections)

## System Overview
- N/A

## Design Considerations

_The system will need to take into consideration a few assumptions ...._

### Assumptions and Dependencies
- At this time, no method inputs or outputs should change
- Async/await will make the code more readable and provide a better developer experience
- Not all promises are candidates for the async/await pattern, but will determine case by case - where the candidate not a good fit, should review for proper handling of error. Exceptions should be documented
- most of the updated are limited to one file, but will evaluate all the files and flag any issues around errorhandling

### General Constraints
- None

### Goals and Guidelines
- Convert promises to await/async where we can
- In either case, make sure all have error handling, and throws a helpful error

### Development Methods
[Code can be found here](https://github.com/retaildevcrews/helium-typescript/blob/master/src/db/cosmosdbprovider.ts#L62)

Here is an example of the proposed change:
```javascript
    public async queryDocuments(query: string): Promise<any[]> {
        // Wrap all functionality in a promise to avoid forcing the caller to use callbacks
        return new Promise(async (resolve, reject) => {
            const { resources: queryResults } = await this.cosmosContainer.items.query(query, this.feedOptions).fetchAll();

            resolve(queryResults);
            reject("Cosmos Error");
        });
    }
```

to 

```javascript
        public async queryDocuments(query: string): Promise<any> {
            try {
                return { resources: queryResults } = await this.cosmosContainer.items.query(query, this.feedOptions).fetchAll();
            } catch {
                return "Cosmos Error";
            }
        }
```

### Developer Friendly
- Using async/await provides a consistent experience when dealing with code in general with predictable way to catch errors

### Reusable/Adaptable
- N/A

### Secure
- N/A

## System Architecture
- N/A

### Design Review Checklist

- [ ] Design Review Meeting Scheduled
- [ ] Design Review approved by team consensus
- [ ] Initial tasks created as issues in main project and assigned to the project backlog with proper priority
- [ ] Test plan complete (unit, validation, smoke)
- [ ] Add "References #<issue number>" to relevant tasks

### PR DESCRIPTION
**Pull request by tobiaswright**
## Type of PR

- [ ] Documentation changes
- [x] Code changes
- [ ] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes

## Purpose of PR
Swap out promises for async/await
## Issues Closed or Referenced

- Closes #78 (this will automatically close the issue when the PR closes)
- References #<issue number> (this references the issue but does not close with PR)

### CODE REVIEW

File: src/db/cosmosdbprovider.ts [hunk 1]
```diff
@@ -52,38 +52,40 @@ export class CosmosDBProvider {
...
      */
     public async getDocument(documentId: string): Promise<any> {
-
-        return new Promise(async (resolve, reject) => {
-            const { resource: result, statusCode: status } =
-                await this.cosmosContainer.item(documentId, QueryUtilities.getPartitionKey(documentId)).read();
+        try {
+            const { resource: result, statusCode: status }
+                = await this.cosmosContainer.item(documentId, QueryUtilities.getPartitionKey(documentId)).read();
             if (status === 200) {
```
Comment (aflinchb) [inline src/db/cosmosdbprovider.ts]: When testing this w/ a query for a non-existing actor or movie id, this now returns an empty actor/movie with no 404 error.  Need to handle when the status code is not 200

Comment (tobiaswright) [reply src/db/cosmosdbprovider.ts]: I've updated this to throw then passed to the catch, logged and throws an error

File: src/secrets/keyvaultprovider.ts [hunk 2]
```diff
@@ -30,19 +30,18 @@ export class KeyVaultProvider {
...
-            });
-
-        return secret;
+        try {
+            const { value: secret } = await this.client.getSecret(name);
+            return secret as string;
+        } catch (err) {
+            if (name === "AppInsightsKey") {
+                this.logger.Trace("App Insights Key not set");
+                return " ";
```
[the remark this answers is not in the document]
Comment (tobiaswright) [reply src/secrets/keyvaultprovider.ts]: Our only options are Trace and Error

File: src/secrets/keyvaultprovider.ts [hunk 3]
```diff
@@ -30,19 +30,18 @@ export class KeyVaultProvider {
...
+            } else {
+                this.logger.Error(Error(), "Unable to find secret " + name + " " + err);
+                throw new Error(`Unable to find secret ${name}`);
+            }
```
Comment (aflinchb) [inline src/secrets/keyvaultprovider.ts]: This is logging twice, one here, and then again in config.t when it calls getSecret - we should put the logging where it makes the most sense.  Probably leave logging in config and then only throw here.

File: src/db/cosmosdbprovider.ts [hunk 4]
```diff
@@ -52,38 +52,45 @@ export class CosmosDBProvider {
     public async initialize() {
 
         this.logger.Trace("Initializing CosmosDB Container");
-        this.cosmosContainer = await this.cosmosClient.database(this.databaseId).container(this.containerId);
+        try {
+            this.cosmosContainer = await this.cosmosClient.database(this.databaseId).container(this.containerId);
+        } catch (err) {
+            this.logger.Error(Error(err), err);
```
Comment (aflinchb) [inline src/db/cosmosdbprovider.ts]: Should this throw the error and not log (similar to the other comments) - lets sync this morning to make sure I didn't miss a decision at the review I had to leave early :) .

File: src/secrets/keyvaultprovider.ts (hunk 3 above)
Comment (tobiaswright) [reply src/secrets/keyvaultprovider.ts]: Makes sense. done!

Comment (tobiaswright) [reply src/secrets/keyvaultprovider.ts]: Actually, I throw the logging at the every catch. Let's discuss to ensure that is consistent across the other files

**Comment by tobiaswright** [comment]
We have some things to work out with throwing but will tackle under a different issue

**Pull request merged**